src/python/pants/engine/internals/selectors.py

```diff
@@ -590,22 +593,50 @@ def __init__(self, *args: Any) -> None:
         self.params = tuple(args)
 
 
+# A specification for how the native engine interacts with @rule coroutines:
+# - coroutines may await on any of `Get`, `MultiGet`, `Effect` or other coroutines decorated with `@rule_helper`.
+# - we will send back a single `Any` or a tuple of `Any` to the coroutine, depending upon the variant of `Get`.
+# - a coroutine will eventually return a single `Any`.
+RuleSend = Union[Any, Tuple[Any, ...]]
+RuleYield = Union[Get, Tuple[Get, ...]]
+RuleCoroutine = Coroutine[RuleYield, RuleSend, Any]
+NativeEngineGeneratorResponse = Union[
+    PyGeneratorResponseGet,
+    PyGeneratorResponseGetMulti,
+    PyGeneratorResponseBreak,
+]
+
+
 def native_engine_generator_send(
-    func, arg
-) -> PyGeneratorResponseGet | PyGeneratorResponseGetMulti | PyGeneratorResponseBreak:
+    rule: RuleCoroutine, arg: RuleSend | None, err: NativeEngineFailure | None
```

Comment:
The arguments here are slightly confusing: a valid value of `RuleSend` is `None`: that's the initial value sent to a generator. So the `RuleSend` `Union` should probably contain `None`.

With that change, it's maybe more obvious that we always either send a value (possibly `None`) _or_ throw an error. So `arg` and `err` could probably merge into a single argument like (replacing `RuleSend`):
```python
RuleInput = Union[
  # The value used to "start" a Generator.
  None,
  # A single value requested by a Get.
  Any,
  # Multiple values requested by a MultiGet.
  Tuple[Any, ...],
  # An exception to be raised in the Generator.
  NativeEngineFailure,
]
```